clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -127,20 +163,29 @@ public ProduceRequest(Struct struct, short version) {
     public Struct toStruct() {
         // Store it in a local variable to protect against concurrent updates
         Map<TopicPartition, MemoryRecords> partitionRecords = partitionRecordsOrFail();
-        Struct struct = new Struct(ApiKeys.PRODUCE.requestSchema(version()));
+        short version = version();
+        Struct struct = new Struct(ApiKeys.PRODUCE.requestSchema(version));
         Map<String, Map<Integer, MemoryRecords>> recordsByTopic = CollectionUtils.groupDataByTopic(partitionRecords);
         struct.set(ACKS_KEY_NAME, acks);
         struct.set(TIMEOUT_KEY_NAME, timeout);
+
+        if (version >= 3)
+            struct.set(TRANSACTIONAL_ID_KEY_NAME, transactionalId);
+
         List<Struct> topicDatas = new ArrayList<>(recordsByTopic.size());
-        for (Map.Entry<String, Map<Integer, MemoryRecords>> entry : recordsByTopic.entrySet()) {
+        for (Map.Entry<String, Map<Integer, MemoryRecords>> topicEntry : recordsByTopic.entrySet()) {
             Struct topicData = struct.instance(TOPIC_DATA_KEY_NAME);
-            topicData.set(TOPIC_KEY_NAME, entry.getKey());
+            topicData.set(TOPIC_KEY_NAME, topicEntry.getKey());
             List<Struct> partitionArray = new ArrayList<>();
-            for (Map.Entry<Integer, MemoryRecords> partitionEntry : entry.getValue().entrySet()) {
+            for (Map.Entry<Integer, MemoryRecords> partitionEntry : topicEntry.getValue().entrySet()) {
                 MemoryRecords records = partitionEntry.getValue();
                 Struct part = topicData.instance(PARTITION_DATA_KEY_NAME)
-                        .set(PARTITION_KEY_NAME, partitionEntry.getKey())
-                        .set(RECORD_SET_KEY_NAME, records);
+                        .set(PARTITION_KEY_NAME, partitionEntry.getKey());
+
+                if (version >= 3)
```

Comment:
I think we should call the `validateRecords` method for all versions. Also, why do we only do it on `toStruct` instead of the constructor?